typed-transitions/src/Protocol/Codec.hs

```diff
@@ -8,17 +8,122 @@
+      Partial response       -> Partial $ response
+        { more = fmap f . more response
+        , end  = fmap f (end response)
+        }
+
+data Choice input m r where
+  Complete :: input -> m r -> Choice input m r
+  Partial  :: Response input (Fold input m r) (m r) -> Choice input m r
+
+-- | Morally an additive conjunction: only one of 'more' or 'end' is allowed
+-- to be used.
+data Response input more end = Response
+  { more :: input -> more
+    -- ^ If there is more input.
+  , end  :: end
+    -- ^ If there's no more input.
+  }
+
+hoistFold
+  :: forall input m n r .
+     ( Functor n )
+  => (forall x . m x -> n x)
+  -> Fold input m r
+  -> Fold input n r
+hoistFold nat fold = Fold $ fmap recurse (nat (runFold fold))
+  where
+  recurse :: Choice input m r -> Choice input n r
+  recurse choice = case choice of
+    Complete leftover term -> Complete leftover $ nat term
+    Partial response       -> Partial $ response
+      { more = hoistFold nat . more response
+      , end  = nat (end response)
+      }
+
+newtype Input input m = Input
+  { runInput :: m (Maybe (input, Input input m))
+  }
+
+drainInput :: Monad m => Input input m -> m [input]
+drainInput inp = runInput inp >>= \it -> case it of
+  Nothing        -> pure []
+  Just (i, inp') -> fmap ((:) i) (drainInput inp')
+
+noInput :: Applicative m => Input input m
+noInput = Input $ pure Nothing
+
+singletonInput :: Applicative m => input -> Input input m
+singletonInput inp = Input $ pure $ Just $ (inp, noInput)
+
+prependInput :: Applicative m => [input] -> Input input m -> Input input m
+prependInput inp tail = case inp of
+  []      -> tail
+  (i : is) -> Input $ pure (Just (i, prependInput is tail))
+
+-- | Use an 'Input' to run a 'Fold'.
+-- Notice how, no matter what, the output of the fold is given. That's
+-- a great thing to have: it eliminates the case in which a decoder gives
+-- a partial parse even after the end of the input stream.
+--
+-- Any _unconsumed_ input is given back. Unconsumed means the monadic thing
+-- which defines it was never run. It could still be exhausted.
+-- In other words: it's 'Just' whenever the fold finished before the input
+-- was exhausted. It's 'Nothing' whenever the input was exhausted before
+-- the fold chose to finish ('Complete').
+foldOverInput
+  :: ( Monad m )
+  => Fold input m r
+  -> Input input m
+  -> m (r, Maybe (Input input m))
+foldOverInput fold input = runFold fold >>= \choice -> case choice of
+  Complete leftovers it -> flip (,) (Just (prependInput [leftovers] input)) <$> it
+  Partial step -> runInput input >>= \it -> case it of
+    Nothing -> flip (,) Nothing <$> end step
+    Just (i, input') -> foldOverInput (more step i) input'
+
+-- Why bother with all this? Well, it makes using a decoder easier: we can
+-- just pass all of the input and we know we'll get a result, no possibility
+-- of a partial!
+
+type Decoder fail input m r = Fold [input] m (Either fail r)
```

Comment:
Could you explain why `Decoder` is getting lists of inputs at a time rather than a single input.  Is this for efficient decoding of chunks of lazy `ByteString`?

Reply:
I would  move the `Decoder` type alias just after `Encoder`.  Then they would be in one logical unit.

Reply:
> Could you explain why Decoder is getting lists of inputs at a time rather than a single input. Is this for efficient decoding of chunks of lazy ByteString?

When defining the "identity" codecs, like `Ouroboros.Network.Protocol.ChainSync.Codec.Id`, in which the concrete representation is `SomeTransition (ChainSyncMessage header point)`, we need to come up with leftover input whenever the `Fold` finishes. If the concrete type of the `Fold` really is a `SomeTransition ...`, then we can't possibly do this.

One option is to use `Maybe (SomeTransition ...)` as the input and leftover type, another is to use `[SomeTransition ...]`. Of course, We could have `Decoder fail input m r = Fold input m (Either fail r)` and put the `[]` or `Maybe` on as needed. But I think wrapping the input in `[]` as part of the `Decoder` type really is a sensible thing to do: the input can always come in parts, even if the concrete type itself is not a monoid. If it is a monoid, then you can just `mconcat` it if need be.